sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -602,17 +606,11 @@ object ColumnPruning extends Rule[LogicalPlan] {
       p.copy(child = w.copy(
         windowExpressions = w.windowExpressions.filter(p.references.contains)))
 
-    // Eliminate no-op Window
-    case w: Window if w.windowExpressions.isEmpty => w.child
-
-    // Eliminate no-op Projects
-    case p @ Project(_, child) if child.sameOutput(p) => child
-
     // Can't prune the columns on LeafNode
     case p @ Project(_, _: LeafNode) => p
 
     // for all other logical plans that inherits the output from it's children
-    case p @ Project(_, child) =>
+    case p @ Project(_, child) if !child.isInstanceOf[Project] =>
```

Comment:
in case the child is a project, shall we anyway update it with `c.output.filter(allReferences.contains)`? I mean can we instead update the `prunedChild` method to check if `c` is a `Project` and behave accordingly?

Reply:
We already handled project over project at L542

Reply:
yes, I see, makes sense, thanks.

Reply:
This requires a code comment. I believe the others will ask the same q when we read the problem again.